src/com/ms/silverking/log/Log.java

```diff
@@ -19,428 +18,445 @@
+  }
+
+  public static void log(Level level, String m) {
+    if (alertReceiver != null && level.intValue() >= alertLevel) {
+      alertReceiver.sendAlert(new Alert(alertContext, level.intValue(), alertKey != null ? alertKey : m, m, alertData));
+    }
+    logDest.log(level, m);
+  }
+
+  public static void warning(String m, Object o) {
+    if (levelMet(Level.WARNING)) {
+      warning(m + o);
+    }
+  }
+
+  public static void warning(Object o) {
+    if (o == null) {
+      warning("null");
+    } else {
+      warning(o.toString());
+    }
+  }
+
+  public static void warning(String m) {
+    log(Level.WARNING, m);
+  }
+
+  public static void warningAsyncf(String f, Object... args) {
+    if (levelMet(Level.WARNING)) {
+      warningAsyncInternal(new LogEntry(Level.WARNING, String.format(f, args)));
+    }
+  }
+
+  public static void warningf(String f, Object... args) {
+    if (levelMet(Level.WARNING)) {
+      warning(String.format(f, args));
+    }
+  }
+
+  public static void warningAsync(String m, Object o) {
+    if (levelMet(Level.WARNING)) {
+      warningAsyncInternal(new LogEntry(Level.WARNING, o, m));
+    }
+  }
+
+  public static void warningAsync(Object o) {
+    if (o == null) {
+      warningAsync("null");
+    } else {
+      warningAsync(o.toString());
+    }
+  }
+
+  public static void warningAsync(String m) {
+    if (levelMet(Level.WARNING)) {
+      warningAsyncInternal(new LogEntry(Level.WARNING, m));
+    }
+  }
+
+  private static void warningAsyncInternal(LogEntry le) {
+    try {
+      logQueue.put(le);
+    } catch (InterruptedException ie) {
+    }
+  }
+
+  public static void severe(String m, String className, String methodName) {
+    log(Level.SEVERE, className + "." + methodName + ": " + m);
+  }
+
+  public static void severe(String m, Object o) {
+    if (levelMet(Level.SEVERE)) {
+      severe(m + o);
+    }
+  }
+
+  public static void severe(Object o) {
+    if (o == null) {
+      severe("null");
+    } else {
+      severe(o.toString());
+    }
+  }
+
+  public static void severe(String m) {
+    log(Level.SEVERE, m);
+  }
+
+  public static void severeAsyncf(String f, Object... args) {
+    if (levelMet(Level.SEVERE)) {
+      severeAsync(String.format(f, args));
+    }
+  }
+
+  public static void severef(String f, Object... args) {
+    if (levelMet(Level.SEVERE)) {
+      severe(String.format(f, args));
+    }
+  }
+
+  public static void severeAsync(String m, Object o) {
+    if (levelMet(Level.SEVERE)) {
+      try {
+        logQueue.put(new LogEntry(Level.SEVERE, o, m));
+      } catch (InterruptedException ie) {
+      }
+    }
+  }
+
+  public static void severeAsync(Object o) {
+    if (o == null) {
+      severeAsync("null");
+    } else {
+      severeAsync(o.toString());
+    }
+  }
+
+  public static void severeAsync(String m) {
+    if (levelMet(Level.SEVERE)) {
+      try {
+        logQueue.put(new LogEntry(Level.SEVERE, m));
+      } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)